### PR TITLE
Add a KYC datatype

### PR DESCRIPTION
**Pull request by charrsky**
#239

### CODE REVIEW

**Review by echatav — CHANGES_REQUESTED** [review]

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 1]
```diff
@@ -355,3 +356,11 @@ instance (Symbolic c, NumberOfBits (BaseField c) ~ n) => Iso (FieldElement c) (B
...
 instance (Symbolic c, NumberOfBits (BaseField c) ~ n) => Iso (ByteString n c) (FieldElement c) where
   from (ByteString a) = fromBinary a
+
+instance (FromConstant Natural (ByteString 8 c), Concat (ByteString 8 c) (ByteString n c))
+    => FromJSON (ByteString n c) where
+    parseJSON = Haskell.fmap fromString . parseJSON @String
+
+instance (ToConstant (ByteString n (Interpreter (Zp p))))
+    => ToJSON (ByteString n (Interpreter (Zp p))) where
+    toJSON = toJSON . toConstant
```
[the remark this answers is not in the document]
Comment (charrsky) [reply src/ZkFold/Symbolic/Data/ByteString.hs:366]: well, I'm not sure how to convert our symbolic bytestring into just string, so I left it as is for now

File: src/ZkFold/Symbolic/Data/KYC.hs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+data KYCData context = KYCData
+    { kycType  :: KYCByteString context
+    , kycValue :: KYCByteString context
+    , kycHash  :: KYCByteString context
+    , kycID    :: UInt 64 Auto context
+    } deriving Generic
+
+isCitizen :: (Symbolic c) => KYCByteString c -> Vector n (KYCByteString c) -> Bool c
+isCitizen citizenship allowedCountries = foldl' (||) false (parFmap (== citizenship) allowedCountries)
```
[the remark this answers is not in the document]
Comment (charrsky) [reply src/ZkFold/Symbolic/Data/KYC.hs]: I tried but then got some type error. Just replaced my foldl'-definition with `ZkFold.Symbolic.Data.Bool`'s any, thanks!

File: src/ZkFold/Symbolic/Data/ByteString.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (echatav) [reply src/ZkFold/Symbolic/Data/ByteString.hs:366]: We must maintain the invariant property `Aeson.decode . Aeson.encode = Just`

**Review by echatav — CHANGES_REQUESTED** [review]
Verify invariant JSON property.

**Comment by TurtlePU** [comment]
@charrsky, for resolving the issue @echatav mantioned, you can convert from/to JSON via Haskell's `ByteString` type. It is `base64`-encoded and should work properly.

File: src/ZkFold/Symbolic/Data/KYC.hs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+data KYCData context = KYCData
+    { kycType  :: KYCByteString context
+    , kycValue :: KYCByteString context
+    , kycHash  :: KYCByteString context
+    , kycID    :: UInt 64 Auto context
+    } deriving Generic
+
+isCitizen :: (Symbolic c) => KYCByteString c -> Vector n (KYCByteString c) -> Bool c
+isCitizen citizenship allowedCountries = any (== citizenship)
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/KYC.hs:24]: How does this pass the typechecker? `any` is not applied to any `Foldable` as far as I can see
I guess `allowedCountries` argument is to be removed

File: src/ZkFold/Symbolic/Data/ByteString.hs [hunk 4]
```diff
@@ -355,3 +357,11 @@ instance (Symbolic c, NumberOfBits (BaseField c) ~ n) => Iso (FieldElement c) (B
 
 instance (Symbolic c, NumberOfBits (BaseField c) ~ n) => Iso (ByteString n c) (FieldElement c) where
   from (ByteString a) = fromBinary a
+
+instance (FromConstant Natural (ByteString 8 c), Concat (ByteString 8 c) (ByteString n c))
```
Comment (TurtlePU) [inline src/ZkFold/Symbolic/Data/ByteString.hs:361]: Much better, but now we still need some tests to make sure this is right
I'd propose to add a test to `Tests/ByteString.hs` module which checks that `parseJSON . toJSON = Just`

Comment (echatav) [reply src/ZkFold/Symbolic/Data/ByteString.hs:361]: Please also add same JSON property test but for the `UInt`s too. These are good tests to have for these basic symbolic types.

**Review by echatav — APPROVED** [review]
JSON invariance property for a couple base symbolic types is now tested

File: symbolic-base/src/ZkFold/Symbolic/Data/KYC.hs
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Symbolic/Data/KYC.hs:1]: @vlasin, should this be in `symbolic-base` or somewhere else?

**Review by TurtlePU — APPROVED** [review]

File: symbolic-base/src/ZkFold/Symbolic/Data/KYC.hs
Comment (vlasin) [reply symbolic-base/src/ZkFold/Symbolic/Data/KYC.hs:1]: I propose to create a `symbolic-apps` package that will contain Symbolic code for this and our other side-projects.

**Pull request merged**